### PR TITLE
SI-7622 Clean Up Phase Assembly

### PR DESCRIPTION
**Pull request by som-snytt**
Allow phases (internal or external) to opt-out of assembly. In particular, plug-ins and optimizer phases that are not enabled will not be plugged in.

Otherwise, this PR is not a clean up so much as a hosing down the sidewalk. "Phase assembly" itself deserves closer inspection.

I think this is still useful, maybe someone agrees.  The fun part is mostly the -Xshow-phases, to see what's enabled or not.

This yet-another-rebase supersedes https://github.com/scala/scala/pull/2790.

### CODE REVIEW

**Comment by gkossakowski** [comment]
@som-snytt: I've got two commits left to review but so far this looks like great work.

I'll try to finish reviewing tomorrow as it's getting late in my time zone.

**Comment by som-snytt** [comment]
#partest is the #ifdiff feature, if it's a feature and not an anti-feature, for defining check files that cope with differential output.  It tests a baked-in flag like java7 or avian or now it can check flags like -optimise (-optimize).

On "right after": OK.  I recall I consciously wanted to shift from "right after" to "follow", because "right after" lacks gravitas, but if people are used to it, no problem.  It reminded me of the "enteringPhase" nomenclature issue.

Thanks for looking.  I thought it collided with your plans, so I was prepared to shelve it, then I reconsidered.

**Comment by som-snytt** [comment]
@adriaanm Help!  I modified IcodeTest together with its clients in this old commit.  (Old in github years.)

IIRC, the issue is that it has to handle whether -optimise is enabled.

Since the partest partition needs another tweak anyway, could these PRs happen simultaneously?

In future, it would be a Par-Test PR followed by the adaptive PR in the scala tests.

(For simplicity, I'm zapping two debug lines in partest that also conflicted.)

**Comment by dragos** [comment]
There are a couple of IDE tests related to loading the continuations plugin, and they fail on this pull request. It might be due to the way we test that the plugin was loaded: `comp.plugins` :) Do we now need to pass `-Pconitnuations:enable` in order to get it in `comp.plugins`?

```
Failed tests:   continuationsPluginIsAlwaysLoaded(scala.tools.eclipse.compiler.settings.ContinuationPluginSettingsTest): Loaded plugins:  expected:<List(continuations)> but was:<List()>
  loadContinuationsPluginVia_XpluginsdirCompilerSetting(scala.tools.eclipse.compiler.settings.ContinuationPluginSettingsTest): Loaded plugins:  expected:<List(continuations)> but was:<List()>
  loadContinuationsPluginVia_XpluginCompilerSetting(scala.tools.eclipse.compiler.settings.ContinuationPluginSettingsTest): Loaded plugins:  expected:<List(continuations)> but was:<List()>
```

**Comment by som-snytt** [comment]
@dragos Yes, in this regime, continuationsPluginIsNotAlwaysLoaded.

@adriaanm The failure is for the icode test submitted to partest. There is no magic involved; I could add the test code on scala side and migrate it later if that's better.  In future, that might be the model: enhance some test infrastructure locally and then push it upstream, just like in really independent projects.

**Comment by adriaanm** [comment]
@som-snytt happy to help, but I'm having trouble reconstructing the comments in their historical context (historical in PR years). In general, test should pass here, with preliminary infrastructure work being done first in the modules, where necessary -- happy to rev the module version counter where needed

**Comment by som-snytt** [comment]
Updated per @gkossakowski guidance.  Added IcodeComparison with extra-special scaladoc.  (The use case is that you need to specify a phase to write icode for both unoptimized and optimized set-up, since the optimizer phase will be absent in the first case.)

[Ideal workflow for this situation: move the broken test to pending, flagged as depending on the next partest bump, then when the bump happens, the tests are automatically restored. (Or some other mechanism to disable a test. Maybe the same mechanism as skip.)]

File: src/partest-extras/scala/tools/partest/IcodeComparison.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/* NSC -- new Scala compiler
```
Comment (adriaanm) [inline src/partest-extras/scala/tools/partest/IcodeComparison.scala:1]: any comments vis-a-vis https://github.com/scala/scala-partest/pull/2/files#L0L18?

Comment (adriaanm) [inline src/partest-extras/scala/tools/partest/IcodeComparison.scala:1]: hey, that looks like LOL18

Comment (som-snytt) [inline src/partest-extras/scala/tools/partest/IcodeComparison.scala:1]: Then I could kill that, LOL.

**Comment by adriaanm** [comment]
> @dragos Yes, in this regime, continuationsPluginIsNotAlwaysLoaded.

This is going to to be the case soon, anyway, as we plan to deprecate the CPS plugin and build it as a separate module (see other thread).

**Comment by som-snytt** [comment]
@dragos @adriaanm Should I try to fix something on the IDE integration?  This PR won't change anybody's life but it's liable to go stale if it sits awhile.

When continuations is purely optional, you might like to know -Xshow-phases that it's available but not enabled.

File: src/compiler/scala/tools/nsc/Global.scala [hunk 2]
```diff
@@ -452,9 +454,9 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
   object patmat extends {
     val global: Global.this.type = Global.this
     val runsAfter = List("typer")
-    // patmat doesn't need to be right after typer, as long as we run before supperaccesors
-    // (sbt does need to run right after typer, so don't conflict)
     val runsRightAfter = None
+    // patmat doesn't need to be right after typer, as long as we run before superaccessors
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/Global.scala:458]: hey, who accessed my supper!?

**Comment by gkossakowski** [comment]
@dragos: how can we proceed with this? I'd like this PR to be merged as it cleans up places in a compiler that I wanted to be clarified for a long time now.

**Comment by VladUreche** [comment]
As per @som-snytt's request on internals, I had a quick look at this PR and it LGTM. It won't do everything we need for plugins, but it's a good start. :+1:

**Comment by gkossakowski** [comment]
@dragos: ping

**Comment by dragos** [comment]
Sorry for going silent. It's still not clear to me why other tests fail, other than "continuations is always loaded". We'll have a look and get back.

**Comment by gkossakowski** [comment]
@som-snytt: FYI, the IDE team fixed problems on their end here: scala-ide/scala-ide#531. Once that PR is merged I believe we'll be able to merge yours.

**Comment by som-snytt** [comment]
@gkossakowski Thanks. I admire the IDE team, in part because all those Eclipse identifiers are like twenty chars long.  Actually, I guess `forceEnableContinuationForNewerScalaVersion` is more like forty.  I vow to help make this change worthy of forty chars.

**Comment by adriaanm** [comment]
Looks like scala-ide/scala-ide#531 was merged!

**Comment by gkossakowski** [comment]
@som-snytt: I just realized that I missed the addition of `IcodeComparison`. Can you explain why it's needed and why does it belong to `partest-extras` and not just partest?

**Comment by som-snytt** [comment]
I lacked on guidance on making a migrating change, where partest gets a change that breaks downstream tests. (That explains the addition, rather than edit IcodeTest.)

Also, way back, there was no guidance on turn-around for pushing new test infrastructure downstream. In fact, this was right before the partest split and it got unlucky on the timing.

I'm willing to be a guinea pig on that workflow, even if it means pushing this PR out to 2.12.

My suggestion would be to follow forking workflow, where you fork something and then contribute upstream, rather than contribute upstream and wait for it to float down to where you can snag it before it goes over the waterfall.

File: src/compiler/scala/tools/nsc/Global.scala (hunk 2 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/Global.scala:458]: Must have been the guy I saw eating verison.

**Comment by gkossakowski** [comment]
@som-snytt: there's no need to delay this PR. I agree that _temporary_ forking (and code duplication) is fine. Could you:
- [x] add a TODO comment to `IcodeComparison` that this class should be moved to partest and removed from `partest-extra`
- [x] file an issue against partest to include that class

Once that is done I'll go ahead and merge this PR.

**Comment by som-snytt** [comment]
@gkossakowski Done. But what if the partest guy rejects our PR?  Does anybody know where he lives?

**Comment by gkossakowski** [comment]
@som-snytt: we'll figure it out.

**Comment by gkossakowski** [comment]
LGTM.

Issue created here: https://issues.scala-lang.org/browse/SI-7836

I'm merging it. Thanks @som-snytt for being patient when it comes to this PR.

**Pull request merged**